src/java.base/share/classes/com/sun/crypto/provider/ConstructKeys.java

```diff
@@ -161,29 +158,42 @@ private static final PrivateKey constructPrivateKey(byte[] encodedKey,
      * @return a secret key constructed from the encodedKey.
      */
     private static final SecretKey constructSecretKey(byte[] encodedKey,
-                                              String encodedKeyAlgorithm)
-    {
-        return (new SecretKeySpec(encodedKey, encodedKeyAlgorithm));
+            int ofs, int len, String encodedKeyAlgorithm) {
+        return (new SecretKeySpec(encodedKey, ofs, len, encodedKeyAlgorithm));
     }
 
     static final Key constructKey(byte[] encoding, String keyAlgorithm,
-                                  int keyType)
-        throws InvalidKeyException, NoSuchAlgorithmException {
-        Key result = null;
+            int keyType) throws InvalidKeyException, NoSuchAlgorithmException {
         switch (keyType) {
         case Cipher.SECRET_KEY:
-            result = ConstructKeys.constructSecretKey(encoding,
-                                                      keyAlgorithm);
-            break;
+            return ConstructKeys.constructSecretKey(encoding, 0,
+                    encoding.length, keyAlgorithm);
         case Cipher.PRIVATE_KEY:
-            result = ConstructKeys.constructPrivateKey(encoding,
-                                                       keyAlgorithm);
-            break;
+            return ConstructKeys.constructPrivateKey(encoding, keyAlgorithm);
         case Cipher.PUBLIC_KEY:
-            result = ConstructKeys.constructPublicKey(encoding,
-                                                      keyAlgorithm);
-            break;
+            return ConstructKeys.constructPublicKey(encoding, keyAlgorithm);
+        default:
+            throw new RuntimeException("Unsupported key type");
```

Comment:
Good improvement to thrown an exception rather than return "null"!

Is NoSuchAlgorithmException fitted better the specification?  See also the comment in  KeyWrapCipher.engineUnwrap().

Reply:
Yes, will change.